### PR TITLE
Reintroduce support for NET40

### LINKED ISSUE #1452: Again support .NET 4.0 for WinForms and WPF
### Again support .NET 4.0 for WinForms and WPF

As discussed by email, I believe really good libraries should not be limited to the latest bells and whistles, as not all applications can keep up with Microsoft release cycles and support deadlines. .NET 4.0 is the last version of .NET supported on WinXP, and a tool e.g. used on an autonomous machine in a production environment may be tied to the OS version used on that machine. So I ask for maintaining support for .NET 4.0 for WinForms and WPF, even knowing that Microsoft officially ended support for .NET 4.0. (Note that if the Microsoft dates were strictly followed, .NET 3.5 SP1 would still have to be supported.  .NET 3.5 SP1 is officially supported until 2023/2028.)

### PR DESCRIPTION
**Pull request by objorke**
Fixes #1452 .

### Checklist

- [ ] I have included examples or tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file
- [x] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- Reintroduce support for NET40 for the core library, WindowsForms and WPF

@oxyplot/admins

### CODE REVIEW

**Comment by VisualMelon** [comment]
Huh... I had issues with [`IReadOnlyList<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlylist-1?view=netframework-4.8) not existing in 4.0 in [my branch](https://github.com/VisualMelon/oxyplot/tree/WinformsNet40Support). I'll submit those changes separately, since there is yet another bug in the histogram helpers that needs addressing.

File: Source/OxyPlot/Utilities/TypeExtensions.cs [hunk 1]
```diff
@@ -26,12 +26,20 @@ public static class TypeExtensions
         /// <returns>An object that represents the specified property, or null if the property is not found.</returns>
         public static PropertyInfo GetRuntimeProperty(this Type type, string name)
         {
+#if NET40
+            var source = type.GetProperties();
```
Comment (VisualMelon) [inline Source/OxyPlot/Utilities/TypeExtensions.cs]: I think this should be `var source = type.GetProperties(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance);` to be equivalent to `var source = typeInfo.AsType().GetRuntimeProperties();`

File: Source/OxyPlot/PlotModel/PlotElement.cs [hunk 2]
```diff
@@ -147,10 +147,15 @@ protected CultureInfo ActualCulture
         /// <remarks>This method creates the hash code by reflecting the value of all public properties.</remarks>
         public virtual int GetElementHashCode()
         {
+#if NET40
+            var properties = this.GetType().GetProperties(BindingFlags.Public|BindingFlags.Instance);
```
Comment (VisualMelon) [inline Source/OxyPlot/PlotModel/PlotElement.cs]: Formatting: spaces around the `|`?

File: Source/OxyPlot/Foundation/CodeGenerator/CodeGenerator.cs [hunk 3]
```diff
@@ -137,7 +137,11 @@ public string ToCode()
         private string Add(object obj)
         {
             var type = obj.GetType();
+#if NET40
+            var hasParameterLessCtor = type.GetConstructors().Any(c => c.GetParameters().Length == 0);
```
Comment (VisualMelon) [inline Source/OxyPlot/Foundation/CodeGenerator/CodeGenerator.cs]: `DeclaredConstructors` retrieves static constructors, which `Type.GetConstructors` does not (without suitable `BindingFlags`). This code probably shouldn't be picking up static constructors at all, so I think the NET4.0 code is fine, but the NET Standard code needs an additional check so it only counts non-static constructors.

Example of counting static constructors:

     typeof(System.Collections.Generic.List<int>).GetTypeInfo().DeclaredConstructors.Count(ci => ci.GetParameters().Length == 0) // returns 2

Comment (objorke) [reply Source/OxyPlot/Foundation/CodeGenerator/CodeGenerator.cs]: Thanks! It is likely to be other bugs like that...

I think OxyPlot.Tests and ExampleLibrary should also target net40, so we could run all tests on all framework versions...

**Pull request merged**